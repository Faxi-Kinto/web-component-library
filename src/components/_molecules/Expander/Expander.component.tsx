import React, { ReactNode, useState } from 'react';
import * as Styled from './Expander.styles';
import { Button } from '../../..';

export type ExpanderProps = {
  title?: string;
  body?: React.ReactNode;
  icon: JSX.Element;
  iconClassName?: string;
  headerClassName?: string;
  buttonClassName?: string;
  bodyClassName?: string;
  headerColor?: string;
  headerSize?: string;
  textColor?: string;
};

const Expander: React.FC<ExpanderProps> = (
  props: ExpanderProps
): JSX.Element => {
  const {
    title,
    body,
    icon,
    iconClassName,
    headerClassName,
    buttonClassName,
    headerColor = '#4E606A',
    headerSize = '18px',
    textColor = '#4E606A',
  } = props;

  const [isShown, setShowHide] = useState(false);

  const handleExpander = () => {
    setShowHide(!isShown);
  };

  return (
    <Styled.Container
      className="expander-card"
      headerColor={headerColor}
      headerSize={headerSize}
      textColor={textColor}
    >
      <div
        className={`expander-card__header${
          headerClassName ? ' ' + headerClassName : ''
        }`}
      >
        <Button
          secondary
          className={`expander-card__header__button${
            buttonClassName ? ' ' + buttonClassName : ''
          }`}
          onClick={handleExpander}
        >
          <span>{title}</span>
          <div className="expander-card__header__wrapper">
            {icon &&
              React.cloneElement(icon, {
                className: `expander-card__header__wrapper__icon ${iconClassName}${
                  isShown ? ' expander-card__header__wrapper__icon--open' : ''
                }`,
              })}
          </div>
        </Button>
      </div>
      <div
        className={`expander-card__box${
          isShown
            ? ' expander-card__box--open'
            : ' expander__card-header__body--close'
        }`}
      >
        <div className="expander-card__box__body">{body}</div>
      </div>
    </Styled.Container>
  );
};

export default Expander;
