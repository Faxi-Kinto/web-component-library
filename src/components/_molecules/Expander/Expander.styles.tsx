import styled from 'styled-components';
import { pxToRem } from '@faxi/web-css-utilities';

export type ExpanderStylingProps = {
  headerColor?: string;
  headerSize?: string;
  textColor?: string;
};

export const Container = styled.details<ExpanderStylingProps>`
  width: 100%;
  overflow: hidden;

  &.expander {
    &-enter {
      height: var(--first);
    }

    &-enter-active {
      height: var(--last);
      transition: height 300ms;
    }

    &-exit {
      height: var(--first);
    }

    &-exit-active {
      height: var(--last);
      transition: height 300ms;
    }
  }

  > summary {
    cursor: pointer;
    user-select: none;
    list-style-type: none;

    &:focus {
      outline: none;
    }

    &::marker {
      display: none;
    }

    &::-webkit-details-marker {
      display: none;
    }
  }

  > p {
    margin: 1rem 0 0;
  }

  .expander {
    &__content {
      display: flex;
      align-items: center;
    }

    &__icon {
      display: inline-block;
      margin-left: ${pxToRem('12px')};
      transition: transform 300ms;

      &--open {
        transform: rotateZ(180deg);
      }
    }
  }
`;
