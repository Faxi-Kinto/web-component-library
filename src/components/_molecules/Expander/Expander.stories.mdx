import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import Expander from './';

<Meta title="Core|Molecules/Expander" />

# Expander Component

## Implementation

#### Default

<Preview>
  <Story name="default">
    <Expander
      title="Explanation"
      body="This graph shows the number of users who have successfully joined your community. 
    The higher the number, the easier it is for people to find someone else to share their journeys with. 
    Increase it by inviting more people or talk to your account manager about how you can promote your community."
      icon={
        <img
          style={{ width: '14px' }}
          src="assets/icons/angle-down.svg"
          alt=""
        />
      }
    />
  </Story>
</Preview>

  <Story name="another">
    <Expander
      title="Explanation"
      body="This graph shows the number of users who have successfully joined your community. 
    The higher the number, the easier it is for people to find someone else to share their journeys with. 
    The higher the number, the easier it is for people to find someone else to share their journeys with. 
    The higher the number, the easier it is for people to find someone else to share their journeys with. 
    The higher the number, the easier it is for people to find someone else to share their journeys with. 
    Increase it by inviting more people or talk to your account manager about how you can promote your community."
      icon={
        <img
          style={{ width: '14px' }}
          src="assets/icons/angle-down.svg"
          alt=""
        />
      }
    />
  </Story>

  <Story name="another one">
    <Expander
      title="Explanation"
      body="This graph shows the number of users who have successfully joined your community. 
    The higher the number, the easier it is for people to find someone else to share their journeys with."
      icon={
        <img
          style={{ width: '14px' }}
          src="assets/icons/angle-down.svg"
          alt=""
        />
      }
    />
  </Story>

## Props

<Props of={Expander} />
