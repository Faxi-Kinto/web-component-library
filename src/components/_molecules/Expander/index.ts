import Expander from './Expander.component';

export default Expander;
