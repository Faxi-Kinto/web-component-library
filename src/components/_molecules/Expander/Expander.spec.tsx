import React from 'react';
import renderer from 'react-test-renderer';
import Expander from './Expander.component';

test('Expander renders correctly', () => {
  const component = renderer.create(
    <Expander
      title="Explanation"
      body="This graph shows the number of users who have successfully joined your community. 
  The higher the number, the easier it is for people to find someone else to share their journeys with. 
  Increase it by inviting more people or talk to your account manager about how you can promote your community."
      icon={
        <img
          style={{ width: '14px' }}
          src="assets/icons/angle-down.svg"
          alt=""
        />
      }
    />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
