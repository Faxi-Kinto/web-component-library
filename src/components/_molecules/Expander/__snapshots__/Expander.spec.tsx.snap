// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expander renders correctly 1`] = `
.c0 {
  width: 100%;
  overflow: hidden;
}

.c0.expander-enter {
  height: var(--first);
}

.c0.expander-enter-active {
  height: var(--last);
  -webkit-transition: height 300ms;
  transition: height 300ms;
}

.c0.expander-exit {
  height: var(--first);
}

.c0.expander-exit-active {
  height: var(--last);
  -webkit-transition: height 300ms;
  transition: height 300ms;
}

.c0 > summary {
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  list-style-type: none;
}

.c0 > summary:focus {
  outline: none;
}

.c0 > summary::marker {
  display: none;
}

.c0 > summary::-webkit-details-marker {
  display: none;
}

.c0 > p {
  margin: 1rem 0 0;
}

.c0 .expander__content {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .expander__icon {
  display: inline-block;
  margin-left: 0.75rem;
  -webkit-transition: -webkit-transform 300ms;
  -webkit-transition: transform 300ms;
  transition: transform 300ms;
}

.c0 .expander__icon--open {
  -webkit-transform: rotateZ(180deg);
  -ms-transform: rotateZ(180deg);
  transform: rotateZ(180deg);
}

<details
  className="c0 uniq"
>
  <style>
    
          details.uniq { 
            --first:undefinedpx; 
            --last: 0px; 
            }
  </style>
  <summary
    onClick={[Function]}
  >
    <div
      className="expander__content"
    >
      Explanation
      <span
        className="expander__icon"
      >
        <img
          alt=""
          className=""
          src="assets/icons/angle-down.svg"
          style={
            Object {
              "width": "14px",
            }
          }
        />
      </span>
    </div>
  </summary>
  <p>
    This graph shows the number of users who have successfully joined your community. 
  The higher the number, the easier it is for people to find someone else to share their journeys with. 
  Increase it by inviting more people or talk to your account manager about how you can promote your community.
  </p>
</details>
`;
